### PR TITLE
kernel_copy: avoid panic on unexpected OS error

### LINKED ISSUE #91152: Copying multiple large files results in `assertion failed: (left == right)` in kernel_copy.rs:592
I tried this code:

```rust
fn copy_file(path: &Path, target: &str) -> io::Result<()> {
    let mut reader = File::open(path)?;
    let mut writer = File::create(make_target_path(path, target))?;

    io::copy(&mut reader, &mut writer)?;

    Ok(())
}
```

I expected the copy operation to succeed (that is, to return `Ok(())`) or to fail (that is, to return `Err(BUY A BETTER HARD DRIVE!)`. What happens instead is a panic thrown from line 592 of kernel_copy.rs. Before I changed from the default mount options to something else, this happened pretty much every time I tried to copy multiple large files from a Linux laptop to a Linux server on my network.

The server in question has a BTRFS volume. That volume was mounted on the laptop by the Gnome shell--er, that is, just whatever Gnome does to mount shares it finds on the network. Because it gets mounted at something obscene, like `/this/path/takes/forever/to/type1000/100/volumename`, I added a symlink to allow access via `~/share` in my home directory. (I only mention this because I already discovered that a symlink renders something like `fs::copy` unworkable, so.... /shrug)

Anyway, looks like the code in kernel_copy 1) encounters an error, 2) attempts to fall back to user space copy instead of using the fancy kernel copy routine, and 3) panics upon realizing that some data has already been written. I reckon there's some kind of reason for it, but I was still surprised that this is a panic instead of just, "Whoops. We encountered an error."

### Meta

`rustc --version --verbose`:
```
rustc 1.56.1 (59eed8a2a 2021-11-01)
binary: rustc
commit-hash: 59eed8a2aac0230a8b53e89d4e99d55912ba6b35
commit-date: 2021-11-01
host: x86_64-unknown-linux-gnu
release: 1.56.1
LLVM version: 13.0.0
```

### PR DESCRIPTION
**Pull request by birkenfeld**
According to documentation, the listed errnos should only occur
if the `copy_file_range` call cannot be made at all, so the
assert be correct.  However, since in practice file system
drivers (incl. FUSE etc.) can return any errno they want, we
should not panic here.

Fixes #91152

### CODE REVIEW

File: library/std/src/sys/unix/kernel_copy.rs [hunk 1]
```diff
@@ -584,13 +584,19 @@ pub(super) fn copy_regular_files(reader: RawFd, writer: RawFd, max_len: u64) ->
...
+                        // not be returned if something was already written, but they happen in
+                        // the wild, see #91152.)
                         //
                         // ¹ these cases should be detected by the initial probe but we handle them here
                         //   anyway in case syscall interception changes during runtime
                         // ² actually invalid file descriptors would cause this too, but in that case
                         //   the fallback code path is expected to encounter the same error again
-                        assert_eq!(written, 0);
-                        CopyResult::Fallback(0)
+                        if written == 0 {
```
Comment (the8472) [inline library/std/src/sys/unix/kernel_copy.rs:595]: you can lift that condition to the match arm, then the next arm will handle the error case

**Comment by the8472** [comment]
I think this might be an upstream bug or if FUSE is involved then a bug in that code and should be reported in the appropriate place instead of being papered over, no?

The assert simply surfaces something breaking the documented syscall API, so the bug lies there.

File: library/std/src/sys/unix/kernel_copy.rs (hunk 1 above)
Comment (birkenfeld) [reply library/std/src/sys/unix/kernel_copy.rs:595]: will do.

**Comment by birkenfeld** [comment]
The question is how well-defined the syscall API really is. At least the manpage doesn't include wording like "if errno is X, Y or Z, no bytes can have been written", but only says "On error, copy_file_range() returns -1 and errno is set to indicate the error." :man_shrugging:

**Pull request merged**